pmlc/dialect/pxa/transforms/stencil.cc

```diff
@@ -60,21 +60,74 @@ class Stencil {
   BlockArgumentSet accIdxs;
   // All used indices
   BlockArgumentSet allIdxs;
+  // Tensors' order
+  unsigned tensorsOrder[kNumTensors];
+  // M, N, K in inner block
+  mlir::BlockArgument innerIdxs[kNumIndex];
+  // The matrix_idx for the next search
+  unsigned nextMatrixIdx[kNumIndex] = {2, 3, 1};
+  // M, N, K's tiles
+  unsigned tiles[kNumIndex];
+
+  // Even index split only. Specified for each index.
+  bool onlyEven[kNumIndex] = {true, true, true};
 
   // Found Gemm operation data
   MulOperationType mulOpType;
   mlir::AffineLoadOp in1Op;
   mlir::AffineLoadOp in2Op;
   AffineReduceOp outOp;
+  // The best tensors' order
+  unsigned bestTensorsOrder[kNumTensors];
+  // The best index (M, N, K)
+  mlir::BlockArgument bestIdxs[kNumIndex];
+  // The best tiles for (M, N, K)
+  unsigned bestTiles[kNumIndex];
+
+  // The best performance
+  double bestPerf;
+
+  llvm::SmallVector<int64_t, 8> opConstRanges;
 
   void PopulateOpBlockArgumentSet();
   BlockArgumentSet UsedIdxs(unsigned strideInfoIndex);
   void CollectUsedIndices();
   void CollectStrideOneIndices();
   void strideOneIdxs(unsigned indx);
 
+  // Search tensors' order
+  void SearchTensorsOrder();
+  // Search the index, i.e., M, N, K, for the inner block
+  void SearchIndex(unsigned matrix_idx);
+  // Test if idx in tensors[tensor_idx] is stride one index
+  bool IsStrideOne(mlir::BlockArgument idx, unsigned tensor_idx);
+  // Test if idx in the tensors are stride one
+  bool ValidateStrideOne(mlir::BlockArgument idx, unsigned matrix_idx);
+  // Test if idx exists in tensorIdxs[tensor_idx]
+  bool IndexExists(mlir::BlockArgument idx, unsigned tensor_idx);
+  // Test if idx exists in the right place
+  bool ValidateIndexExistance(mlir::BlockArgument idx, unsigned matrix_idx);
+  // For (M, N, K) in the inner block, search their tiles
+  void SearchTiles(unsigned idx);
+
+  int64_t idxRange(mlir::BlockArgument idx);
+
+  // The throughput and startup cost of M*N*K matrix multiplication
+  std::pair<double, unsigned> Throughput(unsigned m, unsigned n, unsigned k);
+  // Evaluate the performance of the current searching state
+  double Evaluate();
+
+  // Set the number of threads to a default value if not set from outside.
+  void SetNumberThreads();
+
+  // Number of threads
+  unsigned numThreads;
+
 public:
-  explicit Stencil(mlir::AffineParallelOp opIn) : op(opIn) {}
+  explicit Stencil(mlir::AffineParallelOp opIn, int numThreadsIn)
```

Comment:
no need to name is something different, just go with `numThreads` for the parameter name (even tho it 'collides' with the member name, C++ will sort it out).

Reply:
Old habits die hard. :) To me it is more clear this way. But sure...